c10/core/ConstantSymNodeImpl.h

```diff
@@ -0,0 +1,92 @@
+#include <c10/core/SymNodeImpl.h>
+#include <c10/util/variant.h>
+
+namespace c10 {
+
+// Unlike other SymNodeImpl, this cannot be "dispatched" conventionally,
+// as it typically needs to defer to another SymNodeImpl
+//
+// Can either represent a bool, int (don't support float yet) this is useful
+// for representing otherwise unrepresentable large negative integer constant.
+template <typename T>
+class ConstantSymNodeImpl : public SymNodeImpl {
+  static_assert(
+      std::is_same<T, int64_t>::value || std::is_same<T, bool>::value,
+      "ConstantSymNodeImpl can only accept int64_t or bool types");
+
+ public:
+  ConstantSymNodeImpl(T val) : value_(val) {}
+
+  bool is_int() override {
+    return std::is_same<T, int64_t>::value;
+  }
+  bool is_bool() override {
+    return std::is_same<T, bool>::value;
+  }
+  bool is_float() override {
+    return false;
+  }
+  int64_t guard_int(const char* file, int64_t line) override {
+    TORCH_CHECK(is_int(), "not an int");
+    return int_();
+  }
+  bool guard_bool(const char* file, int64_t line) override {
+    TORCH_CHECK(is_bool(), "not a bool");
+    return bool_();
+  }
+  double guard_float(const char* file, int64_t line) override {
+    TORCH_CHECK(false, "not a float");
+  }
+  int64_t int_() override {
+    TORCH_CHECK(is_int(), "not an int");
+    return c10::get<int64_t>(value_);
+  }
+  bool bool_() override {
+    TORCH_CHECK(is_bool(), "not a bool");
+    return c10::get<bool>(value_);
+  }
+  bool has_hint() override {
+    return true;
+  }
+
+  // Temporary hack to avoid having to implement multiple dispatch for now
+  // Currently even if we have this method, we still raise an error when we get
+  // to SingletonSymNode::eq since comparing with non-singleton is disallowed.
+  // However, we may change that behavior in the future.
+  c10::SymNode eq(const c10::SymNode& other) override {
+    TORCH_INTERNAL_ASSERT(other->singleton_int().has_value());
+    return other->eq(c10::SymNode(make_intrusive<ConstantSymNodeImpl<int64_t>>(
+        c10::get<int64_t>(value_))));
```

Comment:
Why can't you just pass `*this` here?

Reply:
Because it's expecting a SymNode (intrusive ptr to SymNodeImpl) but `*this` gives us ConstantSymNodeImpl is a subclass of SymNodeImpl. 

Updated to just do intrusive_from_this instead:
```cpp
c10::raw::intrusive_ptr::incref(this);
return other->eq(c10::intrusive_ptr<ConstantSymNodeImpl<T>>::reclaim(this));
```